File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
from collections import deque

dx = [0,0,1,-1]
dy = [1,-1,0,0]

T = int(sys.stdin.readline())

for _ in range(T):
    n, m, k = map(int, sys.stdin.readline().split())
    graph = [[0 for _ in range(m)] for _ in range(n)]
    for _ in range(k):
        x, y = map(int, sys.stdin.readline().split())
        graph[x][y] = 1

    cnt = 0
    def bfs(graph, x, y):
        global cnt
        queue = deque([])
        queue.append([x, y])
        graph[x][y] = 0

        while queue:
            a, b = queue.popleft()
            for i in range(4):
                nx = a + dx[i]
                ny = b + dy[i]
                if nx  < 0 or nx >= n or ny < 0 or ny >= m:
                    continue
                if (graph[nx][ny] == 1):
                    graph[nx][ny] = 0
                    queue.append([nx, ny])

    for i in range(n):
        for j in range(m):
            if (graph[i][j] == 1):
                bfs(graph, i, j)
                cnt += 1

    print(cnt)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lines = [];
let idx = -1;
let total_lines = 1;
let testCase = 0;

rl.on("line", (input) => {
  lines.push(input.trim());
  idx++;
  const [a, b, c] = lines[idx].split(" ").map(Number);
  if (c) {
    total_lines += c + 1;
    testCase += 1;
  }

  if (testCase === Number(lines[0]) && lines.length === total_lines) {
    rl.close();
  }
});

rl.on("close", () => {
  const dx = [0, 0, -1, 1];
  const dy = [1, -1, 0, 0];

  let idx = 0;
  const T = Number(lines[idx++]);

  for (let t = 0; t < T; t++) {
    const [m, n, k] = lines[idx++].split(" ").map(Number);
    const grid = Array.from({ length: n }, () => Array(m).fill(0));

    for (let i = idx; i < idx + k; i++) {
      const [x, y] = lines[i].split(" ").map(Number);
      grid[y][x] = 1;
    }
    idx += k;

    function bfs(grid, x, y) {
      const q = [];
      q.push([x, y]);
      grid[x][y] = 0;
      while (q.length) {
        const [x, y] = q.shift();
        for (let i = 0; i < 4; i++) {
          const nx = dx[i] + x,
            ny = dy[i] + y;
          if (0 <= nx && nx < n && 0 <= ny && ny < m) {
            if (grid[nx][ny] === 1) {
              grid[nx][ny] = 0;
              q.push([nx, ny]);
            }
          }
        }
      }
      return;
    }

    let answer = 0;
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < m; j++) {
        if (grid[i][j] === 1) {
          answer++;
          bfs(grid, i, j);
        }
      }
    }
    console.log(answer);
  }
});
